File: apps/api/connectfour.py
import sys
from enum import Enum
from pprint import pprint
from typing import Tuple, Union, Optional
from copy import deepcopy


class Piece(Enum):
    EMPTY = '.'
    PLAYER = 'X'
    COMPUTER = 'O'


class ConnectFour:
    def __init__(
        self,
        width: int,
        height: int,
        board: Optional[list[list[str]]] = None,
    ):
        self.width = width
        self.height = height
        if board is None:
            self.board = self.create_board(width, height)
        else:
            self.board = board

    def create_board(self, width: int, height: int) -> list[list[str]]:
        board = []
        for _ in range(height):
            row = []
            for _ in range(width):
                row.append(Piece.EMPTY.value)
            board.append(row)
        return board

    def print_board(self):
        pprint([[str(x + 1) for x in range(self.width)]])
        pprint(self.board)

    def get_next_open_row(self, col: int) -> Union[int, None]:
        for i in range(self.height - 1, -1, -1):
            if self.board[i][col] == Piece.EMPTY.value:
                return i
        return None

    def get_valid_moves(self) -> list[int]:
        valid_moves = []
        for col in range(self.width):
            if self.get_next_open_row(col) is None:
                continue
            valid_moves.append(col)
        return valid_moves

    def drop_piece(self, col: int, piece: Piece):
        row = self.get_next_open_row(col)
        if row is None:
            return
        self.board[row][col] = piece.value

    def get_vertical_window(
        self,
        col: int,
        row_start: int,
        row_stop: int,
    ) -> list[str]:
        window = []
        for i in range(row_start, row_stop):
            window.append(self.board[i][col])
        return window

    def get_diagonal_window(
        self,
        col_start: int,
        row_start: int,
        direction: int,
    ) -> list[str]:
        window = []
        for i in range(0, 4):
            window.append(self.board[row_start + i * direction][col_start + i])
        return window

    def window_count(self, piece: Piece, window: list[str]) -> int:
        count = 0
        for window_piece in window:
            if window_piece == piece.value:
                count += 1
        return count

    def window_score(self, piece: Piece, window: list[str]) -> int:
        score = 0
        opponent_piece = Piece.PLAYER
        if piece == Piece.PLAYER:
            opponent_piece = Piece.COMPUTER

        if self.window_count(piece, window) == 4:
            score += 100
        elif self.window_count(piece, window) == 3 and self.window_count(
                Piece.EMPTY, window) == 1:
            score += 5
        elif self.window_count(piece, window) == 2 and self.window_count(
                Piece.EMPTY, window) == 2:
            score += 2

        if self.window_count(opponent_piece,
                             window) == 3 and self.window_count(
                                 Piece.EMPTY, window) == 1:
            score -= 4

        return score

    def evaluate(self, piece: Piece) -> int:
        score = 0

        # Center
        center_col_window = self.get_vertical_window(4, 0, self.height)
        center_count = self.window_count(piece, center_col_window)
        score += center_count * 3

        # Horizontal
        for row in range(self.height):
            for col in range(self.width - 3):
                window = self.board[row][col:col + 4]
                score += self.window_score(piece, window)

        # Vertical
        for col in range(self.width):
            for row in range(self.height - 3):
                window = self.get_vertical_window(col, row, row + 4)
                score += self.window_score(piece, window)

        # Positive Diagonal
        for row in range(self.height - 1, 2, -1):
            for col in range(self.width - 3):
                window = self.get_diagonal_window(col, row, -1)
                score += self.window_score(piece, window)

        # Negative Diagonal
        for row in range(self.height - 3):
            for col in range(self.width - 3):
                window = self.get_diagonal_window(col, row, 1)
                score += self.window_score(piece, window)

        return score

    def is_winning_move(self, piece: Piece) -> bool:
        # Horizontal
        for row in range(self.height):
            for col in range(self.width - 3):
                window = self.board[row][col:col + 4]
                if self.window_count(piece, window) == 4:
                    return True

        # Vertical
        for col in range(self.width):
            for row in range(self.height - 3):
                window = self.get_vertical_window(col, row, row + 4)
                if self.window_count(piece, window) == 4:
                    return True

        # Positive Diagonal
        for row in range(self.height - 1, 2, -1):
            for col in range(self.width - 3):
                window = self.get_diagonal_window(col, row, -1)
                if self.window_count(piece, window) == 4:
                    return True

        # Negative Diagonal
        for row in range(self.height - 3):
            for col in range(self.width - 3):
                window = self.get_diagonal_window(col, row, 1)
                if self.window_count(piece, window) == 4:
                    return True

        return False


def minimax(
    game: ConnectFour,
    depth: int,
    alpha: int,
    beta: int,
    maximizer: Piece,
) -> Tuple[int, int]:
    return_value = sys.maxsize
    if maximizer == Piece.COMPUTER:
        return_value = -sys.maxsize - 1
    player_won = game.is_winning_move(Piece.PLAYER)
    computer_won = game.is_winning_move(Piece.COMPUTER)
    valid_moves = game.get_valid_moves()

    if computer_won:
        return sys.maxsize, -1
    if player_won:
        return -sys.maxsize - 1, -1
    if depth == 0:
        return game.evaluate(Piece.COMPUTER), -1
    if len(valid_moves) == 0:
        return 0, -1

    return_col = valid_moves[0]

    for col in valid_moves:
        new_board = deepcopy(game.board)
        new_game = ConnectFour(game.width, game.height, new_board)
        new_game.drop_piece(col, maximizer)
        next_player = Piece.COMPUTER
        if maximizer == Piece.COMPUTER:
            next_player = Piece.PLAYER
        value, _ = minimax(new_game, depth - 1, alpha, beta, next_player)
        if maximizer == Piece.COMPUTER:
            if value > return_value:
                return_value = value
                return_col = col
            if return_value > alpha:
                alpha = return_value
            if alpha >= beta:
                break
        else:
            if value < return_value:
                return_value = value
                return_col = col
            if return_value < beta:
                beta = value
            if beta <= alpha:
                break

    return return_value, return_col


def play(game: ConnectFour):
    try:
        while True:
            game.print_board()
            valid_moves = game.get_valid_moves()
            try:
                raw_col = input("Enter column number: ")
                col = int(raw_col) - 1
            except ValueError:
                print(f'{raw_col} is not a valid column number')
                continue
            if col not in valid_moves:
                print(f'{col} is not a valid move')
                continue
            game.drop_piece(col, Piece.PLAYER)
            player_won = game.is_winning_move(Piece.PLAYER)
            if player_won:
                print('Player has won!')
                game.print_board()
                break
            game.print_board()
            print('Computer Thinking...')
            val, col = minimax(
                game,
                5,
                -sys.maxsize - 1,
                sys.maxsize,
                Piece.COMPUTER,
            )
            print(val, col)
            game.drop_piece(col, Piece.COMPUTER)
            computer_won = game.is_winning_move(Piece.COMPUTER)
            if computer_won:
                print('Computer has won!')
                game.print_board()
                break
    except KeyboardInterrupt:
        print('\nKeyboard interrupt detected, exiting game...')


if __name__ == '__main__':
    init_game = ConnectFour(9, 7)
    play(init_game)

File: apps/goapi/src/connectfour.go
package main

import (
	"fmt"
	"strings"
)

type Board struct {
	height        int
	width         int
	cells         [][]string
	computerPiece string
	playerPiece   string
	emptyPiece    string
}

func createBoard(boardRows []string, computerPiece string, playerPiece string, emptyPiece string) Board {
	b := Board{
		computerPiece: computerPiece,
		playerPiece:   playerPiece,
		emptyPiece:    emptyPiece,
	}
	b.cells = make([][]string, len(boardRows))
	for i := range boardRows {
		b.cells[i] = strings.Split(boardRows[i], "")
	}
	b.height = len(boardRows)
	b.width = len(boardRows[0])
	return b
}

func (b *Board) getNextOpenRow(col int) (int, error) {
	for i := len(b.cells) - 1; i >= 0; i-- {
		if b.cells[i][col] == b.emptyPiece {
			return i, nil
		}
	}
	return -1, fmt.Errorf("no open row found on column %v", col)
}

func (b *Board) dropPiece(piece string, col int) error {
	row, err := b.getNextOpenRow(col)
	if err != nil {
		return err
	}
	b.cells[row][col] = piece
	return nil
}

func (b *Board) getValidMoves() []int {
	validMoves := make([]int, 0, b.width)
	for i := range b.cells[0] {
		_, err := b.getNextOpenRow(i)
		if err != nil {
			continue
		}
		validMoves = append(validMoves, i)
	}
	return validMoves
}

func (b *Board) evaluate(piece string) int {
	score := 0
	// Center columns
	centerCol := b.width / 2
	centerColWindow := b.getVerticalWindow(centerCol, 0, len(b.cells))
	centerCount := windowCount(piece, centerColWindow)
	score += centerCount * 3

	// Horizontal
	for row := range b.cells {
		for col := 0; col < len(b.cells[row])-3; col++ {
			window := b.cells[row][col : col+4]
			score += b.windowScore(piece, window)
		}
	}

	// Vertical
	for col := range b.cells[0] {
		for row := 0; row < len(b.cells)-3; row++ {
			window := b.getVerticalWindow(col, row, row+4)
			score += b.windowScore(piece, window)
		}
	}

	// Positive Diagonal
	for row := len(b.cells) - 1; row >= 3; row-- {
		for col := 0; col < len(b.cells[row])-3; col++ {
			window := b.getDiagonalWindow(col, row, -1)
			score += b.windowScore(piece, window)
		}
	}

	// Negative Diagonal
	for row := 0; row < len(b.cells)-3; row++ {
		for col := 0; col < len(b.cells[row])-3; col++ {
			window := b.getDiagonalWindow(col, row, 1)
			score += b.windowScore(piece, window)
		}
	}
	return score
}

func (b *Board) winningMove(piece string) bool {
	// Horizontal
	for row := range b.cells {
		for col := 0; col < len(b.cells[row])-3; col++ {
			window := b.cells[row][col : col+4]
			count := windowCount(piece, window)
			if count == 4 {
				return true
			}
		}
	}

	// Vertical
	for col := range b.cells[0] {
		for row := 0; row < len(b.cells)-3; row++ {
			window := b.getVerticalWindow(col, row, row+4)
			count := windowCount(piece, window)
			if count == 4 {
				return true
			}
		}
	}

	// Positive Diagonal
	for row := len(b.cells) - 1; row >= 3; row-- {
		for col := 0; col < len(b.cells[row])-3; col++ {
			window := b.getDiagonalWindow(col, row, -1)
			count := windowCount(piece, window)
			if count == 4 {
				return true
			}
		}
	}

	// Negative Diagonal
	for row := 0; row < len(b.cells)-3; row++ {
		for col := 0; col < len(b.cells[row])-3; col++ {
			window := b.getDiagonalWindow(col, row, 1)
			count := windowCount(piece, window)
			if count == 4 {
				return true
			}
		}
	}
	return false
}

func (b *Board) getVerticalWindow(col int, rowStart int, rowStop int) []string {
	window := make([]string, 0, 4)
	for i := rowStart; i < rowStop; i++ {
		window = append(window, b.cells[i][col])
	}
	return window
}

func (b *Board) getDiagonalWindow(colStart int, rowStart int, direction int) []string {
	window := make([]string, 0, 4)
	for i := 0; i < 4; i++ {
		window = append(window, b.cells[rowStart+i*direction][colStart+i])
	}
	return window
}

func (b *Board) copyBoard() Board {
	cb := Board{
		height:        b.height,
		width:         b.width,
		computerPiece: b.computerPiece,
		playerPiece:   b.playerPiece,
		emptyPiece:    b.emptyPiece,
	}
	cb.cells = make([][]string, len(b.cells))
	for i, row := range b.cells {
		rowCopy := make([]string, len(row))
		copy(rowCopy, row)
		cb.cells[i] = rowCopy
	}
	return cb
}

func windowCount(piece string, window []string) int {
	count := 0
	for _, p := range window {
		if piece == p {
			count++
		}
	}
	return count
}

func (b *Board) windowScore(piece string, window []string) int {
	score := 0
	playerPiece := b.computerPiece
	if piece == b.computerPiece {
		playerPiece = b.playerPiece
	}
	if windowCount(piece, window) == 4 {
		score += 100
	} else if windowCount(piece, window) == 3 && windowCount(b.emptyPiece, window) == 1 {
		score += 5
	} else if windowCount(piece, window) == 2 && windowCount(b.emptyPiece, window) == 2 {
		score += 2
	}

	if windowCount(playerPiece, window) == 3 && windowCount(b.emptyPiece, window) == 1 {
		score -= 4
	}
	return score
}
